### PR TITLE
Improve performance when sending large messages

### PR DESCRIPTION
**Pull request by raubitsj**
Description
-----------
When sending and receiving large messages over the service socket we were doing suboptimal handing of the data.

Part of the solution is to avoid copying each buffer everytime a new chunk is received.  The offending code was:
```python
self._data += data
```

The new code path instead keeps track of the receive buffers in a list and only assembles the list when they are needed.  The new logic is contained in a SockBuffer class.

Testing
-------
Create a bunch of files:
```bash
#! /bin/bash
set -e
mkdir -p data/
cd data/
for n in `seq 600000`; do
    dd if=/dev/urandom of=file$( printf %08d "$n" ).bin bs=1 count=$(( (RANDOM % 255) + 1 )) 2>/dev/null
    if [ $(( $n % 1000 )) == 0 ]; then
        echo "$n"
    fi
done
```

Run cvp's artifact reference script pointing to this data directory.
I could not reproduce problems or a great speedup with this change on my machine.   600k files takes ~40 seconds to load.  Is that good?

### CODE REVIEW

File: wandb/sdk/lib/sock_client.py [hunk 1]
```diff
@@ -149,20 +153,48 @@ def send_record_publish(self, record: "pb.Record") -> None:
         server_req.record_publish.CopyFrom(record)
         self.send_server_request(server_req)
 
+    def _buffer_get(self, start: int, end: int, advance: bool = False) -> bytes:
+        if advance:
+            data = b"".join(self._buffer_list)
+            requested = data[start:end]
+            leftover = data[end:]
```
Comment (dmitryduev) [inline wandb/sdk/lib/sock_client.py:160]: rm intermediate step to reduce memory consumption?

File: wandb/sdk/lib/sock_client.py [hunk 2]
```diff
@@ -149,20 +153,48 @@ def send_record_publish(self, record: "pb.Record") -> None:
         server_req.record_publish.CopyFrom(record)
         self.send_server_request(server_req)
 
+    def _buffer_get(self, start: int, end: int, advance: bool = False) -> bytes:
+        if advance:
+            data = b"".join(self._buffer_list)
```
Comment (dmitryduev) [inline wandb/sdk/lib/sock_client.py:158]: once done with it, explicitly del it to help garbage collector?

File: wandb/sdk/lib/sock_client.py (hunk 1 above)
Comment (raubitsj) [reply wandb/sdk/lib/sock_client.py:160]: yeah, the if advance is not fully otpimized.. the fall throguh case is, trying to figure out how to unify them

File: wandb/sdk/lib/sock_client.py (hunk 2 above)
Comment (raubitsj) [reply wandb/sdk/lib/sock_client.py:158]: i just didnt want to deal with the maintenance of the buffer_list once partial data was extracted from it.. but that is what is needed.

**Review by dmitryduev — APPROVED** [review]
🔥

File: wandb/sdk/lib/sock_client.py [hunk 3]
```diff
@@ -55,14 +114,17 @@ def _sendall_with_error_handle(self, data: bytes) -> None:
...
+        while total_sent < total_data:
             start_time = time.monotonic()
             try:
-                sent = self._sock.send(data[total_sent:])
+                sent = self._sock.send(data)
                 # sent equal to 0 indicates a closed socket
                 if sent == 0:
                     raise SockClientClosedError("socket connection broken")
                 total_sent += sent
+                # truncate our data to save memory
```
Comment (kptkin) [inline wandb/sdk/lib/sock_client.py:126]: ah nice optimization!

**Pull request merged**